testing/base/source_gen.cpp

```diff
@@ -0,0 +1,676 @@
+      llvm::is_contained(NonCarbonCppKeywords, id) ||
+      (llvm::is_contained({'i', 'u', 'f'}, id[0]) &&
+       llvm::all_of(id.substr(1),
+                    [](const char c) { return llvm::isDigit(c); })));
+  CARBON_CHECK(id == id_result);
+  CARBON_CHECK(static_cast<int>(id.size()) == length);
+  return id_result;
+}
+
+// Appends a number of unique, random identifiers with a particular length to
+// the provided destination vector.
+//
+// Uses, and when necessary grows, a cached sequence of random identifiers with
+// the specified length. Because these are cached, this is efficient to call
+// repeatedly, but will not produce a different sequence of identifiers.
+auto SourceGen::AppendUniqueIdentifiers(
+    int length, int number, llvm::SmallVectorImpl<llvm::StringRef>& dest)
+    -> void {
+  auto& [count, unique_ids] = unique_ids_by_length.Insert(length, {}).value();
+
+  // See if we need to grow our pool of unique identifiers with the requested
+  // length.
+  if (count < number) {
+    // We'll need to insert exactly the requested new unique identifiers. All
+    // our other inserts will find an existing entry.
+    unique_ids.GrowForInsertCount(count - number);
+
+    // Generate the needed number of identifiers.
+    for ([[maybe_unused]] int i : llvm::seq<int>(count, number)) {
+      // Allocate stable storage for the identifier so we can form stable
+      // `StringRef`s to it.
+      char* id_storage = reinterpret_cast<char*>(
+          storage.Allocate(/*Size=*/length, /*Alignment=*/1));
+      // Repeatedly generate novel identifiers of this length until we find a
+      // new unique one.
+      for (;;) {
+        std::string new_id_tmp = GenerateRandomIdentifier(length);
+        // Copy the new identifier into our stable storage before trying to
+        // insert it into the set so the inserted string ref is stable.
+        memcpy(id_storage, new_id_tmp.data(), length);
+        llvm::StringRef new_id(id_storage, length);
+        auto result = unique_ids.Insert(new_id);
+        if (result.is_inserted()) {
+          break;
+        }
+      }
+    }
+    count = number;
+  }
+  // Append all the identifiers directly out of the set. We make no guarantees
+  // about the relative order so we just use the non-deterministic order of the
+  // set and avoid additional storage.
+  //
+  // TODO: It's awkward the `ForEach` here can't early-exit. This just walks the
+  // whole set which is harmless if inefficient. We should add early exiting
+  // the loop support to `Set` and update this code.
+  unique_ids.ForEach([&](llvm::StringRef id) {
+    if (number > 0) {
+      dest.push_back(id);
+      --number;
+    }
+  });
+  CARBON_CHECK(number == 0);
+}
+
+// Returns a shuffled sequence of integers in the range [min, max].
+//
+// These order of the returned integers is random, but the distribution is
+// computed by appending numbers in round robin over the range up to the
+// requested number.
```

Comment:
I didn't understand what this comment meant until I read the algorithm. Maybe we could rephrase, eg something like:
```suggestion
// The order of the returned integers is random, but each integer in the range appears the same number of times in the result, with the number of appearances rounded up for lower numbers and rounded down for higher numbers in order to exactly produce `number` results.
```